common/workunit/workunit.cpp

```diff
@@ -2657,85 +2660,187 @@ cost_type aggregateCost(const IConstWorkUnit * wu, const char *scope, bool exclu
+IStatisticCollection * GlobalStatisticCollection::getCollectionForUpdate(StatisticCreatorType creatorType, const char * creator, unsigned wfid, const char *graphName, unsigned sgId, bool clearStats)
+{
+    bool wasCreated;
+
+    StatsScopeId graphScopeId;
+    verifyex(graphScopeId.setScopeText(graphName));
+    StatsScopeId wfScopeId(SSTworkflow, wfid);
+    StatsScopeId sgScopeId(SSTsubgraph, sgId);
+
+    IStatisticCollection * sgScopeCollection = statsCollection->ensureSubScopePath({wfScopeId,graphScopeId, sgScopeId}, wasCreated);
+    if (clearStats && !wasCreated)
+        sgScopeCollection->clearStats();
+    sgScopeCollection->markDirty();
+    return createRootStatisticCollection(creatorType, creator, wfScopeId, graphScopeId, sgScopeCollection);
+}
+
+// Re-calculated aggregates from subgraph scope
+bool GlobalStatisticCollection::refreshAggregates()
+{
+    CRuntimeStatisticCollection totals(aggregateKindsMapping);
+    Owned<IBitSet> totalUpdated = createBitSet(aggregateKindsMapping.numStatistics());
+    return statsCollection->refreshAggregates(totals, *totalUpdated);
+}
+
+// Load aggregates in collection from global stats
+void GlobalStatisticCollection::updateAggregates(IWorkUnit *wu)
+{
+    class StatisticsAggregatesWriter : implements IStatisticVisitor
+    {
+        const StatisticsMapping & aggregateKindsMapping;
+        const unsigned numStats;
+        Linked<IWorkUnit> wu;
+    public:
+        StatisticsAggregatesWriter(IWorkUnit * _wu, const StatisticsMapping & _aggregateKindsMapping): wu(_wu), aggregateKindsMapping(_aggregateKindsMapping), numStats(aggregateKindsMapping.numStatistics()) {}
+
+        virtual bool visitScope(const IStatisticCollection & cur)
         {
-            it->next();
+            switch (cur.queryScopeType())
+            {
+            case SSTglobal:
+            case SSTworkflow:
+            case SSTgraph:
+                for (unsigned i=0; i<numStats; ++i)
+                {
+                    StatisticKind kind = aggregateKindsMapping.getKind(i);
+                    stat_type value;
+                    if (cur.getStatistic(kind, value) && value)
+                    {
+                        StringBuffer s;
+                        wu->setStatistic(queryStatisticsComponentType(), queryStatisticsComponentName(), cur.queryScopeType(), cur.getFullScope(s).str(), kind, nullptr, value, 1, 0, StatsMergeReplace);
+                    }
+                }
+                if (cur.queryScopeType()==SSTgraph)
+                    return false;
+                else
+                    return true;
+            default:
+                return false;
+            }
         }
+    };
+    if (refreshAggregates()) // Only serialize if the aggregates has changed
+    {
+        StatisticsAggregatesWriter statsAggregatorWriter(wu, aggregateKindsMapping);
+        statsCollection->visit(statsAggregatorWriter);
```

Comment:
related to the includeStatisticIfZero comment, should this be doing something like : ?

```
Owned<IStatisticGatherer> globalStaatsGatherer = createGlobalStatisticGatherer(wu);
statsCollection->recordStatistics(globalStatsGatherer);
```

CRuntimeStatisticCollection::recordStatistics handles things like includeStatisticIfZero, merge mode and measure conversion.
createGlobalStatisticGatherer builds up a scope as it's descending.

Reply:
statsAggregatorWriter is filtering out the required stats and writing out only those stats. It filters out some global stats that have been loaded in global stats such as dfu stats, compile stats etc.  Stats at the subgraph level should also be ignored (not written to global wu stats).

Also, IStatisticCollection/CStatisticCollection doesn't have a recordStatistics function.